cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/AbstractContext.java

```diff
@@ -290,19 +300,47 @@ public DynamicDatasetCache getDatasetCache() {
     return getDataset(namespace, name, arguments, AccessType.UNKNOWN);
   }
 
-  protected <T extends Dataset> T getDataset(String namespace, String name, Map<String, String> arguments,
-                                             AccessType accessType) throws DatasetInstantiationException {
+  protected <T extends Dataset> T getDataset(final String namespace, final String name,
+                                             final Map<String, String> arguments,
+                                             final AccessType accessType) throws DatasetInstantiationException {
     if (NamespaceId.SYSTEM.getNamespace().equalsIgnoreCase(namespace)) {
       throw new DatasetInstantiationException(String.format("Dataset %s cannot be instantiated from %s namespace. " +
                                                               "Cannot access %s namespace.",
                                                             name, NamespaceId.SYSTEM, NamespaceId.SYSTEM));
     }
-    return datasetCache.getDataset(namespace, name, arguments, accessType);
+
+    try {
+      return Retries.callWithRetries(new Callable<T>() {
+        @Override
+        public T call() throws DatasetInstantiationException {
+          return datasetCache.getDataset(namespace, name, arguments, accessType);
+        }
+      }, retryStrategy);
+    } catch (ExecutionException e) {
+      Throwables.propagateIfInstanceOf(e.getCause(), DatasetInstantiationException.class);
+      throw new DatasetInstantiationException("Error instantiating Dataset " + name, e.getCause());
+    } catch (InterruptedException e) {
+      throw new DatasetInstantiationException("Interrupted while instantiating Dataset " + name, e);
```

Comment:
Here, too, I wonder whether we should unwrap a RetriesExhaustedException. Because the contract of getDataset() is that it throw DatasetInstantiationException, not RetriesExhaustedException.